pkg/testing/matrix/matrix.go

```diff
@@ -0,0 +1,457 @@
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/cmdutil"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/workspace"
+	"github.com/stretchr/testify/assert"
+
+	javagen "github.com/pulumi/pulumi-java/pkg/codegen/java"
+	yamlgen "github.com/pulumi/pulumi-yaml/pkg/pulumiyaml/codegen"
+	dotnetgen "github.com/pulumi/pulumi/pkg/v3/codegen/dotnet"
+	gogen "github.com/pulumi/pulumi/pkg/v3/codegen/go"
+	jsgen "github.com/pulumi/pulumi/pkg/v3/codegen/nodejs"
+	pygen "github.com/pulumi/pulumi/pkg/v3/codegen/python"
+)
+
+const NODEJS = "nodejs"
+const PYTHON = "python"
+const JAVA = "java"
+const DOTNET = "dotnet"
+const YAML = "yaml"
+const GO = "go"
+
+type LangTestOption struct {
+	Language string
+	Opts     *i.ProgramTestOptions
+}
+
+type PluginOptions struct {
+	Name    string
+	Kind    workspace.PluginKind
+	Version semver.Version
+
+	// Build is the set of commands to run to build the plugin and SDK
+	// In the future we want to just generate the SDK using the plugin.
+	Build []exec.Cmd
+
+	// Bin is the path to the plugin to use.
+	Bin string
+}
+
+type TestOptions struct {
+	Languages  []LangTestOption
+	Program    *i.ProgramTestOptions
+	PulumiSDKs map[string]string
+	Plugins    []PluginOptions
+}
+
+type projectGeneratorFunc func(directory string, project workspace.Project, p *pcl.Program,
+	localProjects map[string]string) error
+
+func Test(t *testing.T, opts TestOptions) {
+
+	dir := opts.Program.Dir
+	if !filepath.IsAbs(dir) {
+		pwd, err := os.Getwd()
+		assert.NoError(t, err)
+		dir = filepath.Join(pwd, dir)
+	}
+
+	projfile := filepath.Join(dir, workspace.ProjectFile+".yaml")
+	proj, err := workspace.LoadProject(projfile)
+	assert.NoError(t, err)
+	projinfo := &engine.Projinfo{Proj: proj, Root: dir}
+
+	assert.NoError(t, err)
+
+	//Assert runtime is YAML
+	assert.Equal(t, projinfo.Proj.Runtime.Name(), "yaml")
+
+	pwd, _, ctx, err := engine.ProjectInfoContext(projinfo, nil, cmdutil.Diag(), cmdutil.Diag(), false, nil)
+	assert.NoError(t, err)
+
+	var pclProgram *pcl.Program
+
+	//check if *.pp file exists in dir
+	files, err := ioutil.ReadDir(dir)
+	assert.NoError(t, err)
+	for _, file := range files {
+		if strings.HasSuffix(file.Name(), ".pp") {
+			parser := syntax.NewParser()
+			//get file content
+			ppFilePath := filepath.Join(dir, file.Name())
+			contents, err := ioutil.ReadFile(ppFilePath)
+			assert.NoError(t, err)
+			err = parser.ParseFile(bytes.NewReader(contents), ppFilePath)
+			assert.NoError(t, err)
+			program, diags, err := pcl.BindProgram(parser.Files, pcl.PluginHost(ctx.Host))
+			assert.NoError(t, err)
+			assert.Empty(t, diags)
+			pclProgram = program
+
+		}
+	}
+	if pclProgram == nil {
+		loader := schema.NewPluginLoader(ctx.Host)
+		_, pclProgram, err = yamlgen.Eject(pwd, loader)
+		assert.NoError(t, err)
+	}
+	assert.NotNil(t, proj)
+	assert.NotNil(t, pclProgram)
+
+	localProjects := map[string]string{}
+
+	//Here we are overriding the plugin links to use our own plugin links.
```

Comment:
Assert it's empty before setting it?

Reply:
The issue is that we need to have plugin links in the base YAML sometimes otherwise it can't be read into the program.

Reply:
Right, should we not maintain those links though?

Reply:
Ah, figured out how to use the specified plugins in code to load from YAML. So now plugins should only be specified in one place.